library/std/src/sys/thread_local/fast_local.rs

```diff
@@ -237,8 +237,9 @@ unsafe extern "C" fn destroy_value<T>(ptr: *mut u8) {
     // Wrap the call in a catch to ensure unwinding is caught in the event
     // a panic takes place in a destructor.
     if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| unsafe {
-        let value = (*ptr).inner.take();
-        (*ptr).dtor_state.set(DtorState::RunningOrHasRun);
+        let Key { inner, dtor_state } = &*ptr;
+        let value = inner.take();
+        dtor_state.set(DtorState::RunningOrHasRun);
         drop(value);
```

Comment:
something that's not clear to me is whether we can simply rearrange this: do `dtor_state.set(DtorState::RunningOrHasRun)` and then call the more lightweight `drop_in_place`?

Reply:
We could, and there are some other improvements possible as well, see #116123.